functions/src/generate-sessions-speakers-schedule.js

```diff
@@ -0,0 +1,90 @@
+import * as functions from 'firebase-functions';
+import { firestore } from 'firebase-admin';
+import mapSessionsSpeakersSchedule from './schedule-generator/speakers-sessions-schedule-map'
+import mapSessionsSpeakers from './schedule-generator/speakers-sessions-map'
+
+export const sessionsWrite = functions.firestore.document('sessions/{sessionId}').onWrite( async (change, context) => {
+    return generateAndSaveData();
+});
+
+export const scheduleWrite = functions.firestore.document('schedule/{scheduleId}').onWrite( async (change, context) => {
+    const scheduleConfig = functions.config().schedule;
+    const scheduleEnabled = scheduleConfig && scheduleConfig.enabled === 'true';
+    return scheduleEnabled && generateAndSaveData();
+});
+
+export const speakersWrite = functions.firestore.document('speakers/{speakerId}').onWrite( async (change, context) => {
+    return generateAndSaveData();
+});
+
+async function generateAndSaveData() {
+    const sessionsPromise = firestore().collection('sessions').get();
+    const schedulePromise = firestore().collection('schedule').orderBy('date', 'desc').get();
+    const speakersPromise = firestore().collection('speakers').get();
+
+    const [sessionsSnapshot, scheduleSnapshot, speakersSnapshot] = await Promise.all([sessionsPromise, schedulePromise, speakersPromise]);
+
+    const sessions = {};
+    const schedule = {};
+    const speakers = {};
+
+    sessionsSnapshot.forEach((doc) => {
+        sessions[doc.id] = doc.data();
+    });
+
+    scheduleSnapshot.forEach((doc) => {
+        schedule[doc.id] = doc.data();
+    });
+
+    speakersSnapshot.forEach((doc) => {
+        speakers[doc.id] = doc.data();
+    });
+
+    let generatedData = {}
+    const scheduleConfig = functions.config().schedule;
+    const scheduleEnabled = scheduleConfig && scheduleConfig.enabled === 'true';
+
+    if (!Object.keys(sessions).length) {
+        generatedData = Object.assign({}, speakers);
+    }
+    else if (!scheduleEnabled || !Object.keys(schedule).length) {
+        generatedData = mapSessionsSpeakers(sessions, speakers);
+    }
+    else {
+        generatedData = mapSessionsSpeakersSchedule(sessions, schedule, speakers);
```

Comment:
change the order of params to `sessions, speakers, schedule` to be consistent with the name